File: dataplane/standalone/saiserver/fdb.cc
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "dataplane/standalone/saiserver/fdb.h"

#include <glog/logging.h>

#include "dataplane/proto/sai/common.pb.h"
#include "dataplane/proto/sai/fdb.pb.h"
#include "dataplane/standalone/saiserver/common.h"
#include "dataplane/standalone/saiserver/enum.h"

grpc::Status Fdb::CreateFdbEntry(
    grpc::ServerContext* context,
    const lemming::dataplane::sai::CreateFdbEntryRequest* req,
    lemming::dataplane::sai::CreateFdbEntryResponse* resp) {
  LOG(INFO) << "Func: " << __PRETTY_FUNCTION__;

  return grpc::Status::OK;
}

grpc::Status Fdb::RemoveFdbEntry(
    grpc::ServerContext* context,
    const lemming::dataplane::sai::RemoveFdbEntryRequest* req,
    lemming::dataplane::sai::RemoveFdbEntryResponse* resp) {
  LOG(INFO) << "Func: " << __PRETTY_FUNCTION__;

  grpc::ClientContext context;

  auto status = api->remove_fdb_entry(entry);
  if (!status.ok()) {
    auto it = context.GetServerTrailingMetadata().find("traceparent");
    if (it != context.GetServerTrailingMetadata().end()) {
      LOG(ERROR) << "Lucius RPC error: Trace ID " << it->second
                 << " msg: " << status.error_message();
    } else {
      LOG(ERROR) << "Lucius RPC error: " << status.error_message();
    }
    return grpc::Status::INTERNAL;
  }

  return grpc::Status::OK;
}

grpc::Status Fdb::SetFdbEntryAttribute(
    grpc::ServerContext* context,
    const lemming::dataplane::sai::SetFdbEntryAttributeRequest* req,
    lemming::dataplane::sai::SetFdbEntryAttributeResponse* resp) {
  LOG(INFO) << "Func: " << __PRETTY_FUNCTION__;

  return grpc::Status::OK;
}

grpc::Status Fdb::GetFdbEntryAttribute(
    grpc::ServerContext* context,
    const lemming::dataplane::sai::GetFdbEntryAttributeRequest* req,
    lemming::dataplane::sai::GetFdbEntryAttributeResponse* resp) {
  LOG(INFO) << "Func: " << __PRETTY_FUNCTION__;

  return grpc::Status::OK;
}

grpc::Status Fdb::CreateFdbEntries(
    grpc::ServerContext* context,
    const lemming::dataplane::sai::CreateFdbEntriesRequest* req,
    lemming::dataplane::sai::CreateFdbEntriesResponse* resp) {
  LOG(INFO) << "Func: " << __PRETTY_FUNCTION__;

  return grpc::Status::OK;
}

grpc::Status Fdb::RemoveFdbEntries(
    grpc::ServerContext* context,
    const lemming::dataplane::sai::RemoveFdbEntriesRequest* req,
    lemming::dataplane::sai::RemoveFdbEntriesResponse* resp) {
  LOG(INFO) << "Func: " << __PRETTY_FUNCTION__;

  return grpc::Status::OK;
}

File: dataplane/saiserver/fdb.go
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package saiserver

import (
	"context"
	"fmt"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	"github.com/openconfig/lemming/dataplane/forwarding/fwdconfig"
	saipb "github.com/openconfig/lemming/dataplane/proto/sai"
	"github.com/openconfig/lemming/dataplane/saiserver/attrmgr"
	fwdpb "github.com/openconfig/lemming/proto/forwarding"
)

type fdb struct {
	saipb.UnimplementedFdbServer
	mgr       *attrmgr.AttrMgr
	dataplane switchDataplaneAPI
}

func newFdb(mgr *attrmgr.AttrMgr, dataplane switchDataplaneAPI, s *grpc.Server) (*fdb, error) {
	f := &fdb{
		mgr:       mgr,
		dataplane: dataplane,
	}
	saipb.RegisterFdbServer(s, f)
	return f, nil
}

func (f *fdb) FlushFdbEntries(ctx context.Context, req *saipb.FlushFdbEntriesRequest) (*saipb.FlushFdbEntriesResponse, error) {
	return &saipb.FlushFdbEntriesResponse{}, nil
}

func (f *fdb) CreateFdbEntry(ctx context.Context, req *saipb.CreateFdbEntryRequest) (*saipb.CreateFdbEntryResponse, error) {
	entry := req.GetEntry()
	if entry == nil {
		return nil, status.Errorf(codes.InvalidArgument, "FDB entry is required")
	}

	mac := entry.GetMacAddress()
	if len(mac) == 0 {
		return nil, status.Errorf(codes.InvalidArgument, "MAC address is required")
	}

	portOID := req.GetBridgePortId()
	if portOID == 0 {
		return nil, status.Errorf(codes.InvalidArgument, "Bridge port ID is required")
	}

	bpReq := &saipb.GetBridgePortAttributeRequest{
		Oid:      portOID,
		AttrType: []saipb.BridgePortAttr{saipb.BridgePortAttr_BRIDGE_PORT_ATTR_PORT_ID},
	}
	bpResp := &saipb.GetBridgePortAttributeResponse{}
	if err := f.mgr.PopulateAttributes(bpReq, bpResp); err != nil {
		return nil, fmt.Errorf("failed to populate bridge port %d: %v", portOID, err)
	}

	portID := bpResp.GetAttr().GetPortId()
	if portID == 0 {
		return nil, fmt.Errorf("cannot find port ID for bridge port %d", portOID)
	}

	addReq := fwdconfig.TableEntryAddRequest(f.dataplane.ID(), FDBTable).AppendEntry(
		fwdconfig.EntryDesc(fwdconfig.ExactEntry(
			fwdconfig.PacketFieldBytes(fwdpb.PacketFieldNum_PACKET_FIELD_NUM_ETHER_MAC_DST).WithBytes(mac),
		)),
		fwdconfig.TransmitAction(fmt.Sprint(portID)),
	).Build()

	if _, err := f.dataplane.TableEntryAdd(ctx, addReq); err != nil {
		return nil, fmt.Errorf("failed to add FDB entry to dataplane: %v", err)
	}

	return &saipb.CreateFdbEntryResponse{}, nil
}

func (f *fdb) RemoveFdbEntry(ctx context.Context, req *saipb.RemoveFdbEntryRequest) (*saipb.RemoveFdbEntryResponse, error) {
	entry := req.GetEntry()
	if entry == nil {
		return nil, status.Errorf(codes.InvalidArgument, "FDB entry is required")
	}

	mac := entry.GetMacAddress()
	if len(mac) == 0 {
		return nil, status.Errorf(codes.InvalidArgument, "MAC address is required")
	}

	delReq := fwdconfig.TableEntryRemoveRequest(f.dataplane.ID(), FDBTable).AppendEntry(
		fwdconfig.EntryDesc(fwdconfig.ExactEntry(
			fwdconfig.PacketFieldBytes(fwdpb.PacketFieldNum_PACKET_FIELD_NUM_ETHER_MAC_DST).WithBytes(mac),
		)),
	).Build()

	if _, err := f.dataplane.TableEntryRemove(ctx, delReq); err != nil {
		return nil, fmt.Errorf("failed to remove FDB entry from dataplane: %v", err)
	}

	return &saipb.RemoveFdbEntryResponse{}, nil
}
